src/mtev_http.c

```diff
@@ -1118,66 +1120,88 @@ mtev_http_ctx_acceptor_free(void *v) {
   /* We attempt to consume from the first_input */
   struct bchain *in, *tofree;
+  const char *str_in_f;
   while(1) {
-    const char *str_in_f;
     int rlen;
     in = ctx->req.first_input;
-    if(!in)
+    if(!in) {
       in = ctx->req.first_input = ctx->req.last_input =
-          ALLOC_BCHAIN(DEFAULT_BCHAINSIZE);
-    _fixup_bchain(in);
-    str_in_f = strnstrn("\r\n", 2, in->buff + in->start, in->size);
-    if(str_in_f && ctx->req.next_chunk_read) {
-      if(str_in_f != (in->buff + in->start)) {
-        mtevL(http_debug, "HTTP chunked encoding error, no trailing CRLF.\n");
-        return -1;
+        ALLOC_BCHAIN(DEFAULT_BCHAINSIZE);
+    }
+
+    if (ctx->req.payload_chunked == mtev_false) {
+      if (in->size == in->allocd) {
+        next_chunk = in->size;
+        goto successful_chunk_size;
       }
-      in->start += 2;
-      in->size -= 2;
-      ctx->req.next_chunk_read = 0;
-      ctx->req.next_chunk = 0;
+    } else {
       str_in_f = strnstrn("\r\n", 2, in->buff + in->start, in->size);
-    }
-    if(str_in_f) {
-      unsigned int clen = 0;
-      const char *cp = in->buff + in->start;
-      while(cp <= str_in_f) {
-        if(*cp >= '0' && *cp <= '9') clen = (clen << 4) | (*cp - '0');
-        else if(*cp >= 'a' && *cp <= 'f') clen = (clen << 4) | (*cp - 'a' + 10);
-        else if(*cp >= 'A' && *cp <= 'F') clen = (clen << 4) | (*cp - 'A' + 10);
-        else if(*cp == '\r' && cp[1] == '\n') {
-          mtevL(http_debug, "Found for chunk length(%d)\n", clen);
-          next_chunk = clen;
-          in->start += 2;
-          in->size -= 2;
-          goto successful_chunk_size;
-        }
-        else {
-          mtevL(http_debug, "chunked input encoding error: '%02x'\n", *cp);
-          return -1;
+      if(str_in_f) {
+        unsigned int clen = 0;
+        const char *cp = in->buff + in->start;
+        const char *cp_begin = in->buff + in->start;
+        while(cp <= str_in_f) {
+          if(*cp >= '0' && *cp <= '9') clen = (clen << 4) | (*cp - '0');
+          else if(*cp >= 'a' && *cp <= 'f') clen = (clen << 4) | (*cp - 'a' + 10);
+          else if(*cp >= 'A' && *cp <= 'F') clen = (clen << 4) | (*cp - 'A' + 10);
+          else if(*cp == '\r' && cp[1] == '\n') {
+            mtevL(http_debug, "Found for chunk length(%d)\n", clen);
+            next_chunk = clen;
+            if (in->size - 2 >= clen) {
+              in->start += cp - cp_begin + 2;
+              in->size -= cp - cp_begin + 2;
+              goto successful_chunk_size;
+            } else {
+              /**
+               * we have decoded a chunk length but the current bchain
+               * is not large enough to handle the entire chunk.
+               * 
+               * In this case we allocate a new bchain which is large
+               * enough to hold the entire chunk, copy in the chunk data that
+               * we have already read and then keep reading.
+               */
```

Comment:
I don't think that's true.  If the chunk size (`clen`) was 1MB we would make a 1MB block and read into it.  Only in the case where we had really slow reads would we waste allocations and frees.  If the read after the 1MB alloc did not read 1MB we would do this again.

That code block is not correct as `in->allocd` will always be larger than `clen` for small chunk sizes (32K allocd vs 16K chunk size) and we would never allocate the replacement block.

It might work if you did:

```} else if ((in->allocd - in->start) < clen) { ```

But I don't think we need to worry about this for now.